### PR TITLE
Wire up SELECT DISTINCT

### PR DESCRIPTION
**Pull request by corylanou**
### the raw data

```sql
> select * from names
name: names
-----------
time                    first   last
2015-05-01T00:00:00Z    suzie   smith
2015-05-01T08:00:00Z    frank   smith
2015-05-01T16:00:00Z    jonny   jones
```

### distinct with no where or group by

```sql
> select distinct(first) from names
name: names
-----------
time                    distinct
1970-01-01T00:00:00Z    [frank jonny suzie]

> select distinct(last) from names
name: names
-----------
time                    distinct
1970-01-01T00:00:00Z    [jones smith]
```
###  Distinct with no fields will currently return an error (not supported)

```sql
> select distinct() from names
ERR: error parsing query: distinct function requires at least one argument
```

###  Distinct with more than one field will currently return an error (not supported)

```sql
> select distinct(first, last) from names
ERR: error parsing query: distinct function can only have one argument
```

### Distinct with a group by and no where clause will err out (like all aggregate functions should)

```sql
> select distinct(last) from names group by time(1h)
ERR: error parsing query: aggregate functions with GROUP BY time require a WHERE time clause
```

### Distinct with group by and where should bucket values appropriatly

```sql
> select distinct(last) from names where time >= '2015-05-01T00:00:00Z' and time <= '2015-05-01T16:00:00Z' group by time(12h)
name: names
-----------
time                    distinct
2015-05-01T00:00:00Z    [smith]
2015-05-01T12:00:00Z    [jones]

> select distinct(first) from names where time >= '2015-05-01T00:00:00Z' and time <= '2015-05-01T16:00:00Z' group by time(12h)
name: names
-----------
time                    distinct
2015-05-01T00:00:00Z    [frank suzie]
2015-05-01T12:00:00Z    [jonny]
```

### the raw data

```sql
> select * from cpu
name: cpu
---------
time                    value
2015-05-01T00:00:00Z    1.1
2015-05-01T08:00:00Z    1.2
2015-05-01T16:00:00Z    1.3
2015-05-02T00:00:00Z    2.1
2015-05-02T08:00:00Z    2.2
2015-05-02T16:00:00Z    2.3
2015-05-03T00:00:00Z    3.1
2015-05-03T08:00:00Z    3.2
2015-05-03T16:00:00Z    3.3
2015-05-04T00:00:00Z    4.1
2015-05-04T08:00:00Z    4.2
2015-05-04T16:00:00Z    4.3
```

### distinct with no where or group by

```sql
select distinct(value) from cpu
name: cpu
---------
time                    distinct
1970-01-01T00:00:00Z    [1.1 1.2 1.3 2.1 2.2 2.3 3.1 3.2 3.3 4.1 4.2 4.3]
```

### distinct with different groupings

```sql
> select distinct(value) from cpu where time >= '2015-05-01T00:00:00Z' and time <= '2015-05-04T00:00:00Z' group by time(12h)
name: cpu
---------
time                    distinct
2015-05-01T00:00:00Z    [1.1 1.2]
2015-05-01T12:00:00Z    [1.3]
2015-05-02T00:00:00Z    [2.1 2.2]
2015-05-02T12:00:00Z    [2.3]
2015-05-03T00:00:00Z    [3.1 3.2]
2015-05-03T12:00:00Z    [3.3]
2015-05-04T00:00:00Z    [4.1 4.2]
```

```sql
> select distinct(value) from cpu where time >= '2015-05-01T00:00:00Z' and time <= '2015-05-04T00:00:00Z' group by time(1d)
name: cpu
---------
time                    distinct
2015-05-01T00:00:00Z    [1.1 1.2 1.3]
2015-05-02T00:00:00Z    [2.1 2.2 2.3]
2015-05-03T00:00:00Z    [3.1 3.2 3.3]
2015-05-04T00:00:00Z    [4.1 4.2 4.3]

```

### CODE REVIEW

**Comment by pauldix** [comment]
Looks good so far. Although your sort doesn't handle a `bool` field :).

Curious how you're going to handle it when they pass in a tag name and a `WHERE time` clause ;)

**Comment by corylanou** [comment]
Fixes #1815

File: influxql/ast.go [hunk 1]
```diff
@@ -765,6 +767,32 @@ func (s *SelectStatement) RewriteWildcards(fields Fields, dimensions Dimensions)
...
+		default:
+			rwFields = append(rwFields, f)
+		}
+	}
+	other.Fields = rwFields
+	other.IsRawQuery = false
+
+	return other
+}
+
```
Comment (dgnorton) [inline influxql/ast.go]: I would simplify this method to look like this... https://gist.github.com/dgnorton/3125e459045a5cedc948.

File: influxql/ast.go [hunk 2]
```diff
@@ -900,11 +944,44 @@ func (s *SelectStatement) Validate(tr targetRequirement) error {
...
+			return true
+		}
+	}
+	return false
+}
 
-	if err := s.validateDerivative(); err != nil {
-		return err
+// ValdiateDistinct ensures we have a properly constructed distinct statement
+func (s *SelectStatement) ValidateDistinct() error {
```
Comment (jwilder) [inline influxql/ast.go]: Typo: ValdiateDistinct.  Also does this need to be public?

File: influxql/ast.go [hunk 3]
```diff
@@ -886,6 +914,22 @@ func (s *SelectStatement) hasTimeDimensions(node Node) bool {
...
+	}
+
+	if err := s.validateDerivative(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SelectStatement) ValidateAggregates(tr targetRequirement) error {
```
Comment (jwilder) [inline influxql/ast.go]: Make this private?

File: influxql/ast.go [hunk 4]
```diff
@@ -765,6 +767,32 @@ func (s *SelectStatement) RewriteWildcards(fields Fields, dimensions Dimensions)
...
+	}
+
+	other := s.Clone()
+
+	// Rewrite any `distinct foo` as `distinct(foo)`
+	rwFields := make(Fields, 0, len(s.Fields))
+	for _, f := range s.Fields {
+		switch v := f.Expr.(type) {
+		case *Distinct:
+			field := &Field{Expr: &Call{Name: "distinct", Args: []Expr{&VarRef{Val: v.Val}}}}
```
Comment (jwilder) [inline influxql/ast.go]: Could this be simplified to:
`field := &Field{Expr: v.NewCall()}`

**Comment by dgnorton** [comment]
I agree with @jwilder that the `Validate*` functions should be converted to private...even the pre-existing ones.  Including the root `Validate` function, which is only called at the end of `parseSelectStatement`.

**Comment by dgnorton** [comment]
Need to add more tests for the validation code.  e.g. `select count(distinct(too, many, arguments))` passes.

**Comment by corylanou** [comment]
Addressed all items.  Will address `select count(distinct(too, many, arguments))` tests in #1891 as it has more edge cases to deal with including this one.

**Comment by jwilder** [comment]
LGTM :+1:

**Comment by pauldix** [comment]
Looks good. One thing left over from this is to handle when the user passes in a tag key to distinct. Logged #2612
